### PR TITLE
task 4 Angular. RxJS & Observables. HTTP 

### PR DESCRIPTION
**Pull request by otecpavlo**
1. Task: https://github.com/rolling-scopes-school/tasks/blob/master/tasks/angular/rxjs-observables-http.md
2. Done - 23.08.2021 / deadline 22.08.2021
:heavy_check_mark: Search input debounce is implemented
:heavy_check_mark: Search functionality is integrated with the YouTube API
:heavy_check_mark: Detailed information page uses a call to the YouTube API
:heavy_check_mark: Login block reflects the current login state
:heavy_check_mark: HTTP interceptor is used to pass the token and the base API URL

### CODE REVIEW

File: youtube-client/src/app/auth/services/auth.service.ts [hunk 1]
```diff
@@ -1,26 +1,35 @@
...
   }
 
   logout() {
     localStorage.removeItem('currentUser');
+    this.authState$$.next(false);
     this.router.navigate(['/auth']);
   }
 
-  isLoggedIn(): boolean {
+  isLoggedIn(): string {
```
Comment (4m9fk) [inline youtube-client/src/app/auth/services/auth.service.ts:29]: Why? Oo

File: youtube-client/src/app/core/components/header/login/login.component.ts [hunk 2]
```diff
@@ -7,10 +8,19 @@ import { AuthService } from '../../../../auth/services/auth.service';
   styleUrls: ['./login.component.scss']
 })
 export class LoginComponent {
+  authState?: boolean;
 
-  constructor(private authService: AuthService) { }
+  constructor(private authService: AuthService, private router: Router) {
+    this.authService.authState$.subscribe((response) => {
```
Comment (4m9fk) [inline youtube-client/src/app/core/components/header/login/login.component.ts:14]: We should unsubscribe in components

File: youtube-client/src/assets/config.json [hunk 3]
```diff
@@ -0,0 +1,4 @@
+{
+  "searchOptions": "search?type=video&part=snippet&maxResults=10",
```
Comment (4m9fk) [inline youtube-client/src/assets/config.json:2]: Could we use regular constants file for it?

File: youtube-client/src/app/youtube/common/helpers.ts [hunk 4]
```diff
@@ -0,0 +1,15 @@
+import { IItem } from '../../shared/models/search-response.model';
+
+export const responseHandler = (item: IItem) => ({
```
Comment (4m9fk) [inline youtube-client/src/app/youtube/common/helpers.ts:3]: Please add return type

Comment (4m9fk) [reply youtube-client/src/app/youtube/common/helpers.ts:3]: Also pay more attention to naming

File: youtube-client/src/app/shared/services/http.service.ts [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+import { Injectable } from '@angular/core';
+import { Observable } from 'rxjs';
+import { switchMap } from 'rxjs/operators';
+import { Config } from '../models/config.model';
+import { ISearchResponse, ISearchId } from '../models/search-response.model';
+
+@Injectable({
+  providedIn: 'root'
+})
+export class HttpService {
```
Comment (4m9fk) [inline youtube-client/src/app/shared/services/http.service.ts:11]: Naming: This service is more concrete than just http

File: youtube-client/src/app/core/http-interceptors/auth-interceptor.ts [hunk 6]
```diff
@@ -0,0 +1,21 @@
...
+  HttpEvent, HttpInterceptor, HttpHandler, HttpRequest
+} from '@angular/common/http';
+import { Observable } from 'rxjs';
+
+@Injectable()
+export class AuthInterceptor implements HttpInterceptor {
+
+  intercept(req: HttpRequest<any>, next: HttpHandler):
+  Observable<HttpEvent<any>> {
+    const API_URL = 'https://www.googleapis.com/youtube/v3';
```
Comment (4m9fk) [inline youtube-client/src/app/core/http-interceptors/auth-interceptor.ts:12]: constants should be placed in separate file

File: youtube-client/src/app/auth/services/auth.service.ts (hunk 1 above)
Comment (otecpavlo) [reply youtube-client/src/app/auth/services/auth.service.ts:29]: returned boolean, I don't know why I did that;)

File: youtube-client/src/app/core/components/header/login/login.component.ts (hunk 2 above)
Comment (otecpavlo) [reply youtube-client/src/app/core/components/header/login/login.component.ts:14]: added unsubscribe in ngOnDestroy method

File: youtube-client/src/assets/config.json (hunk 3 above)
Comment (otecpavlo) [reply youtube-client/src/assets/config.json:2]: added `app/common/global-constants.ts`

File: youtube-client/src/app/youtube/common/helpers.ts (hunk 4 above)
Comment (otecpavlo) [reply youtube-client/src/app/youtube/common/helpers.ts:3]: added return type

File: youtube-client/src/app/shared/services/http.service.ts (hunk 5 above)
Comment (otecpavlo) [reply youtube-client/src/app/shared/services/http.service.ts:11]: updated HttpService. It uses http requests only

File: youtube-client/src/app/core/http-interceptors/auth-interceptor.ts (hunk 6 above)
Comment (otecpavlo) [reply youtube-client/src/app/core/http-interceptors/auth-interceptor.ts:12]: added app/common/global-constants.ts

**Pull request merged**